### PR TITLE
CPM-533: Add user intent to set categories on a product

### PR DESCRIPTION
**Pull request by Balamung**
TODO : 
- [x] new validator to make sure categories are all viewable by the user. In CE we have to check the categories exist, in EE we have to check permissions
- [x] update IsUserOwnerOfTheProductValidator to check there will still be owned categories after update
- [x] merge old product categories (that user cannot see) + new categories before update. In EE the NotGrantedCategoryMerger doesn't work during save because the user is not set in the token storage. So it's done by the handler (in a dedicated applier to be precise). The legacy validator doesn't scream if we add a non viewable category

Originally there was one validator to check the categories exist, and one to check the user can see them. But the violation messages are the same. From the user point of view, a non viewable category does not exist. So I put the logic in an unique validator CategoriesShouldBeViewableValidator

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Category/back/API/Query/GetReadCategories.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Category\API\Query;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface GetReadCategories
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Category/back/API/Query/GetReadCategories.php:11]: GetViewabeCategories

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetCategories.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class SetCategories implements CategoryUserIntent
+{
+    /** @param array<string> $categoriesCodes */
+    public function __construct(private array $categoriesCodes)
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetCategories.php:16]: $categoryCodes

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Infrastructure/Validation/CategoriesShouldBeEditableByUserValidatorSpec.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            productIdentifier: 'foo',
+            categoryUserIntent: $categoryUserIntent
+        ));
+        $getOwnedCategories->forUserId(['master'], 1)->willReturn(['master']);
+        $context->buildViolation(Argument::any())->shouldNotBeCalled();
+
+        $this->validate($categoryUserIntent, new CategoriesShouldBeEditableByUser());
+    }
+
+    function it_allows_adding_a_read_access_category(
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Infrastructure/Validation/CategoriesShouldBeEditableByUserValidatorSpec.php:65]: nope, there should be at least one "owned" category

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/CategoriesShouldBeEditableByUserValidator.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        if ([] === $categoryCodes) {
+            return;
+        }
+
+        $ownedCategories = $this->getOwnedCategories->forUserId($categoryCodes, $command->userId());
+        $nonOwnedCategories = \array_values(\array_diff($categoryCodes, $ownedCategories));
+
+        if (\count($nonOwnedCategories) === 0) {
+            return;
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/CategoriesShouldBeEditableByUserValidator.php:45]: no, there should be at least one "owned" category (else the user doesn't "own" the product anymore, and therefore is not permitted to update it)
```suggestion
        $ownedCategories = $this->getOwnedCategories->forUserId($categoryCodes, $command->userId());

        if (\count($ownedCategories) === 0) {
            $this->context->buildViolation('a relevant message')->addViolation();
        }
```

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/CategoriesShouldBeEditableByUserValidator.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+        $readCategories = $this->getReadCategories->forUserId($nonOwnedCategories, $command->userId());
+        $nonReadCategories = \array_values(\array_diff($nonOwnedCategories, $readCategories));
+
+        if (\count($nonReadCategories) > 0) {
+            $this->context->buildViolation($constraint->message)->addViolation();
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/CategoriesShouldBeEditableByUserValidator.php:52]: I would extract the 2 checks in 2 private methods for more readability (e.g
$this->checkUserIsStillOwner();
$this->checkCategoriesAreViewable();

File: src/Akeneo/Pim/Enrichment/Product/back/Application/SetCategoriesApplier.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * (c) 2022 Akeneo SAS (https://www.akeneo.com)
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\Pim\Enrichment\Product\Application;
+
+use Akeneo\Pim\Enrichment\Category\API\Query\GetViewableCategories;
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/SetCategoriesApplier.php:16]: TODO fix the coupling

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Query/SqlGetNonViewableCategoryCodes.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+use Akeneo\Pim\Enrichment\Category\API\Query\GetViewableCategories;
+use Akeneo\Pim\Enrichment\Product\Domain\Query\GetCategoryCodes;
+use Akeneo\Pim\Enrichment\Product\Domain\Query\GetNonViewableCategoryCodes;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class SqlGetNonViewableCategoryCodes implements GetNonViewableCategoryCodes
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Query/SqlGetNonViewableCategoryCodes.php:15]: It's not really "Sql", I don't know how to name it

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/IsUserOwnerOfTheProductValidator.php [hunk 8]
```diff
@@ -37,17 +41,47 @@ public function validate($command, Constraint $constraint): void
...
+
+    private function checkAtLeastOneNewCategoryIsGranted(CategoryUserIntent $categoryUserIntent, int $userId): bool
+    {
+        if ($categoryUserIntent instanceof SetCategories) {
+            $newCategoryCodes = $categoryUserIntent->categoryCodes();
+        } else {
+            throw new \LogicException('Not implemented');
+        }
+
+        if ([] === $newCategoryCodes) {
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/IsUserOwnerOfTheProductValidator.php:80]: Should merge non viewable categories and these $newCategoryCodes

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Query/SqlGetNonViewableCategoryCodes.php (hunk 7 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Query/SqlGetNonViewableCategoryCodes.php:15]: just remove Sql

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php [hunk 9]
```diff
@@ -78,6 +79,124 @@ public function it_creates_a_new_uncategorized_product(): void
...
+        $this->createProduct('my_product', ['categories' => ['print']]);
+        $this->getContainer()->get('pim_catalog.validator.unique_value_set')->reset(); // Needed to update the product
+
+        $this->expectException(ViolationsException::class);
+        $this->expectExceptionMessage('You should at least keep your product in one category on which you have an own permission');
+
+        $command = new UpsertProductCommand(
+            userId: $this->getUserId('betty'),
+            productIdentifier: 'identifier',
+            categoryUserIntent: new SetCategories(['sales']) // betty can view 'sales'category, but is not owner.
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php:157]:
```suggestion
            categoryUserIntent: new SetCategories(['sales']) // betty can view 'sales' category, but is not owner.
```

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php [hunk 10]
```diff
@@ -78,6 +79,124 @@ public function it_creates_a_new_uncategorized_product(): void
         Assert::assertNotNull($product->getValue('name'));
     }
 
+    /** @test */
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php:82]: IsUserOwnerValidator should be in group sequence to stop the validation when the user is not owner
and split this validator:
- is owner of current categories in the group sequence 1
- is still owner after update in last group sequence

**Review by cemorin — APPROVED** [review]

**Review by nmarniesse — APPROVED** [review]

**Pull request merged**